### ISSUE TITLE
Mixer E2E Tests are flaky

### ISSUE THREAD

**Issue by hklai** [istio/istio#9858]
Mixer E2E tests have become flaky lately. Apparently due to:

1. Init problem
Example:
https://circleci.com/gh/istio/istio/240720
https://gubernator.k8s.io/build/istio-prow/logs/e2e-mixer-no_auth/7717

2. Test problem
https://circleci.com/gh/istio/istio/241026
https://gubernator.k8s.io/build/istio-prow/logs/e2e-mixer-no_auth/7784

**Comment by kyessenov** [comment]
Prow test failed due to set-up. Prometheus service node port binding failed for some reason at infrastructure level.

**Comment by hklai** [comment]
@douglas-reid See https://circleci.com/gh/istio/istio/240388 for a similar failure in master.

**Comment by duderino** [comment]
@hklai how does mixer look now?  can this be closed?

**Comment by hklai** [comment]
@duderino still looking flaky
Postsubmit: https://testgrid.k8s.io/istio-postsubmits#e2e-mixer-no_auth
Postsubmit: https://testgrid.k8s.io/istio-presubmits#e2e-mixer-no_auth

**Comment by douglas-reid** [comment]
Unfortunately, a lot of the failures still seem related to init issues:

`2018-11-13T01:28:48.516541Z	error	Failed to complete Init. Error context deadline exceeded`.

**Comment by hklai** [comment]
Another TestCheckCache failure (not init related)
https://circleci.com/gh/istio/istio/257511?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Can we get that fixed first?

**Comment by douglas-reid** [comment]
@hklai I'm fairly certain that was addressed by the Makefile changes in `release-1.1` (#9857), but it doesn't seem like CircleCI is using those changes in that merge PR. Any ideas why not?

**Comment by douglas-reid** [comment]
@hklai sure. do we know why it started lying ?

**Comment by irisdingbj** [comment]
@douglas-reid  I have a PR  https://storage.googleapis.com/istio-prow/pr-logs/pull/istio_istio/10308 which will make sure it use your generate_e2e_test_yaml:

cat install/kubernetes/namespace.yaml > install/kubernetes/istio.yaml
/home/prow/go/out/linux_amd64/release/helm template --set global.tag=19ce4a723282daeb4459783baf863854688e3605 \
	--name=istio \
	--namespace=istio-system \
	--set global.hub=gcr.io/istio-testing \
	--set global.proxy.enableCoreDump=false \
	--set global.proxy.concurrency=1 \
	--set prometheus.scrapeInterval=1s \
	--set gateways.istio-ingressgateway.autoscaleMax=1 \
	--set mixer.policy.replicaCount=2 \
	--set mixer.policy.autoscaleEnabled=false \
	--values install/kubernetes/helm/istio/values.yaml \
	install/kubernetes/helm/istio >> install/kubernetes/istio.yaml
cat install/kubernetes/namespace.yaml > install/kubernetes/istio-auth.yaml
/home/prow/go/out/linux_amd64/release/helm template --set global.tag=19ce4a723282daeb4459783baf863854688e3605 \
	--name=istio \
	--namespace=istio-system \
	--set global.hub=gcr.io/istio-testing \
	--set global.mtls.enabled=true \
	--set prometheus.scrapeInterval=1s \
	--set gateways.istio-ingressgateway.autoscaleMax=1 \
	--set mixer.policy.replicaCount=2 \
	--set mixer.policy.autoscaleEnabled=false \
	--set global.controlPlaneSecurityEnabled=true \
	--set global.proxy.enableCoreDump=false \
	--set global.proxy.concurrency=1 \
	--values install/kubernetes/helm/istio/values.yaml \
	install/kubernetes/helm/istio >> install/kubernetes/istio-auth.yaml
set -o pipefail; go test -v -timeout 35m ./tests/e2e/tests/mixer \
--auth_enable=false --egress=false --ingress=false --rbac_enable=false \

**Comment by irisdingbj** [comment]
@douglas-reid I have submitted a PR https://github.com/istio/istio/pull/10308 which will make sure e2e_mixer use your generate_e2e_test_yaml.

However it the TestCheckCache is still flaky.

**Comment by kyessenov** [comment]
https://github.com/istio/istio/pull/10260 should help with check cache flakiness.

**Comment by douglas-reid** [comment]
Hopefully, @kyessenov 's PR will merge soon. Are there any other flaky tests within this umbrella?

**Comment by hklai** [comment]
Comparing https://testgrid.k8s.io/istio-presubmits#circleci-e2e-mixer-noauth-v1alpha3-v2 and https://testgrid.k8s.io/istio-presubmits#e2e-mixer-no_auth, it looks like the init problem happens more often on circle.

Example: https://circleci.com/gh/istio/istio/265998

I don't suppose that is fixed by #10260, right?

**Comment by kyessenov** [comment]
@hklai The example has failed to deploy istio. Can we please separate the infrastructure/test environment issues from actual issues in istio? #10260 might help reduce flakiness in TestCheckCache. It did have a very low rate of flakes reported since June (per my observation). It probably will not have effect on others.

**Comment by hklai** [comment]
@kyessenov sure

Looking at https://testgrid.k8s.io/istio-postsubmits#circleci-e2e-mixer-noauth-v1alpha3-v2, TestTcpMetrics, TestMetric and TestCheckCache are the top 3 most flaky tests. Have they all been looked at?

From https://circleci.com/gh/istio/istio/265998, it looks like the deployment error was ImagePullBackOff.  do you have any idea why that occasionally happened?

**Comment by kyessenov** [comment]
@hklai 

In my limited experience, metric tests flake due to delay in pulling metrics through prometheus. I'm not sure why that happens, or how can we force faster metrics pull.

Re: image pull -- I would guess something to do with the docker registry being slow/rate limiting.

**Comment by douglas-reid** [comment]
Using the circleci tests to guide flakes based on history is unfortunately poisoned by the fact that circleci wasn't using the install yamls that reduced the prometheus scrape times.

I'm hoping `TestCheckCache` will turn out to be less flaky with that addressed.

I'll try to monitor the others this week.

**Comment by sdake** [comment]
Here is another example of a context deadline failure reported:
https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/10269/e2e-dashboard/9688

this time in e2e-dashboard in a different PR.

**Comment by esnible** [comment]
For #7863 I confirmed that the problem is that at least one CircleCI machine has IPv4 forwarding turned off.  If IPv4 forwarding is turned off the Docker builds fail but the build doesn't stop, leading to image pull failures.

I will track down the specific machines.  We can decide if we want the build to turn IPv4 forwarding back on or if we want to exclude those machines from the build.

**Comment by kyessenov** [comment]
Would stopping the build when the docker push fails be a sufficient workaround for now?

**Comment by esnible** [comment]
@kyessenov Stopping the build when the `docker build` fails would allow the test to fail in 1 minute instead of 13 minutes.  It would still fail.

I have a PR to gather more data at https://github.com/istio/istio/pull/10777 but it hasn't failed yet.  (I have actually never seen it fail in a PR, only in Gubernator reports, which is mysterious.)

The PR is supposed to `exit 1` fail when there is a problem in addition to gathering more information to troubleshoot this.  It might be worth merging this PR to gather more data, although I would prefer if we could verify expected before when forwarding is disabled first.

**Comment by douglas-reid** [comment]
A quick check on this issue:

- the postsubmit job shows that most of the recent failures have not been test failures, but infrastructure issues. For example, https://gubernator.k8s.io/build/istio-prow/logs/e2e-mixer-no_auth/9902: 12 tests passed, 1 test was skipped, and yet the job is marked as a failure (`*** Test killed with quit: ran too long (36m0s).`). This killed when run too long was happening when the test itself was cleaning up and just took too long. This happened on all but one failure (total of 4) since 12/25/2018. I will consider the postsubmit job to no longer be FLAKY for now.
- The presubmit has several failures due to test length as well. For now, I will ignore those. It also has a lot of `Failed to complete Init. Error context deadline exceeded` failures (as discussed above). I will also ignore those. However, there are a few failures that do seem to be real test logic issues: `TestCheckCache`, `TestTcpMetrics`, `TestKubeenvMetrics`, and `TestRedis*Quota`. While these don't seem to have individual high rates of failure, combined they represent a real issue.

I would like to close this issue and open four separate issues (also as 1.1 P1s) for those individual tests to allow more targeted scoping in addressing them.

Thoughts?

**Comment by hklai** [comment]
Would increasingly the timeout be sufficient to address these deadline
exceeded issues?

**Comment by douglas-reid** [comment]
@hklai maybe?

I'm happy to come back to look at those issues once the individual tests are passing consistently. 

Ultimately, I don't think we want tests to run for more than 36min. However, if that is the only failure happening, I'll consider it a short-term success (which hopefully the new framework, etc., will help with).

**Comment by hklai** [comment]
@douglas-reid 

agreed. the timeout is meant for killing hanging tests. If some tests actually need that much time to complete (i.e. statistics don't lie) we should increase the timeout. =)

**Comment by douglas-reid** [comment]
OK. We now have issues opened for the four individual flaky tests in e2e_mixer (all with release-1.1, P1 labels) with some PRs open against them. I'm going to close this issue, with an eye towards opening another issue if/when the long-running issue becomes the blocker.

**Comment by hklai** [comment]
@douglas-reid I have sent https://github.com/istio/istio/pull/10878 to hopefully improve the first problem you mentioned in https://github.com/istio/istio/issues/9858#issuecomment-452120297.
